### ISSUE TITLE
Service Events - Register/Unregister API update

### ISSUE THREAD

**Issue by rolsonquadras** [hyperledger/aries-framework-go#276]
Based on comments in https://github.com/hyperledger/aries-framework-go/pull/240#discussion_r325332114

The current code behaviour is as follows.

1. If consumer doesn't call RegisterEvent or RegisterAutoExecute - service will not handle messages and throw error for incoming messages.
2. If consumer doesn't call RegisterEvent and calls RegisterAutoExecute - service will auto execute the messages.
3. If consumer doesn't call RegisterAutoExecute and calls RegisterEvent - service will trigger the event and wait for callback
4. If RegisterAutoExecute and RegisterEvent both are called - service will behave same as pt#2 or pt#3 based on which was registered last.
5. If RegisterAutoExecute and RegisterEvent both are called AND service calls either UnregisterAutoExecute or  UnregisterEvent - service will behave same as pt#1.
6. If consumer calls RegisterEvent twice - last channel registered will receive the message.
7. If any consumer calls UnregisterEvent - service removes the event irrespective of the consumer

**Comment by rolsonquadras** [comment]
The changes required.
1. Throw error if Register event is called second time with different channel (pt#6)
2. Support to pass channel as argument in UnregisterEvent(pt#7) - throw error if trying to unregistered with different channel than the one used in RegisterEvent
3. If RegisterAutoExecute and RegisterEvent both are called AND service calls either UnregisterAutoExecute or UnregisterEvent -- fallback to  pt#2 or pt#2 based on unregister function call.

**Comment by rolsonquadras** [comment]
@troyronda Please have a look and let me know if anything is missing or incorrect.

**Comment by llorllale** [comment]
Note: this point from the top post doesn't make sense considering pts 1, 2 and 4

> If RegisterAutoExecute and RegisterEvent both are called AND service calls either UnregisterAutoExecute or UnregisterEvent - service will behave same as pt#1.

Aren't these the possible states?

|AutoExecute|RegisterForEvents|Result|
|----------------|--------------|---------------|
|enabled|disabled|execute protocol automatically|
|disabled|enabled|fire events, wait for callbacks|
|disabled|disabled|return error|
|enabled|enabled|**impossible**|

This suggests that calling `RegisterEvent` should also call `UnregisterAutoExecute()` and `RegisterAutoExecute()` should call `UnregisterEvent()`. Note that the `Unregister*` functions should not call their `Register*` counterparts because it looks like we want to keep the disabled/disabled state?

**Comment by rolsonquadras** [comment]
> Note: this point from the top post doesn't make sense considering pts 1, 2 and 4
> 
> > If RegisterAutoExecute and RegisterEvent both are called AND service calls either UnregisterAutoExecute or UnregisterEvent - service will behave same as pt#1.

Yes, this was in pt#3 in my second [post](https://github.com/hyperledger/aries-framework-go/issues/276#issuecomment-532850874). We can discuss about the ideal behaviour for this case.

> 
> Aren't these the possible states?
> 
> AutoExecute	RegisterForEvents	Result
> enabled	disabled	execute protocol automatically
> disabled	enabled	fire events, wait for callbacks
> disabled	disabled	return error
> enabled	enabled	**impossible**
> This suggests that calling `RegisterEvent` should also call `UnregisterAutoExecute()` and `RegisterAutoExecute()` should call `UnregisterEvent()`. Note that the `Unregister*` functions should not call their `Register*` counterparts because it looks like we want to keep the disabled/disabled state (not to mention a nice stackoverflow if we follow my previous suggestion)?

That's an implementation details. From behaviour/requirement point of view, the question is should the service implicitly change the behaviour or should we make the consumer to `UnregisterX` and then call `RegisterY` (X/Y = AutoExecute/Event).

**Comment by llorllale** [comment]
@rolsonquadras looks like you and I are on the same page: we're trying to make sense of all scenarios here and pick the ones we want to implement.

I attempted to make sense of it all in my previous comment. I suggest we map them out as such to improve clarity because it is getting hard to follow what's going on.

**Comment by rolsonquadras** [comment]
Added diagram for current and proposed behaviour. 

**Current Behaviour:**

![Service_Event_Current](https://user-images.githubusercontent.com/10111768/65277508-0f015180-daf8-11e9-94a6-21c43c6678c6.png)

**Proposed Behaviour:**

![Service_Event_Proposed](https://user-images.githubusercontent.com/10111768/65277509-0f015180-daf8-11e9-9606-5bd13e0225d1.png)

**Comment by rolsonquadras** [comment]
@troyronda @llorllale any feedback on the proposed behaviour ?

**Comment by troyronda** [comment]
- Are you proposing using Deregister rather than Unregister? I assume this was a typo?
- It appears that you are proposing two functions that do the same thing. UnregisterEvent vs UnregisterAutoexecute. Are you planning to have a difference?

**Comment by troyronda** [comment]
I would consider replacing the AutoExecute mechanism with a utility function that simply calls the callbacks.

The auto execute APIs are a bit of a wart, IMO.

**Comment by rolsonquadras** [comment]
@troyronda @llorllale As per our discussion and above input, I have updated the proposed design flow. 

![Service_Event_Proposed_v2](https://user-images.githubusercontent.com/10111768/65440362-a0bdd700-ddf6-11e9-93f8-c946c63e7a6e.png)
